### ISSUE TITLE
Ability to prepopulate merge vs rebase choice based on project policy preference

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-egit/egit#2]
I work on Eclipse/JDT - the project policies are to opt for Rebase when pulling (Right click on repo, choose Pull ...) or when I create a branch and configure upstream for push/pull.

Most of the time when working with egit, these dialogs come pre-populated with Merge. Given how divergent the course of action these choices implement, I am paranoid about the wrong default choice being picked up.

This issue is to request a way to configure and control the pre-poplulation choice of Rebase v Merge.

(If there is already a way to do that, apologies for the noise - I did ask around among more expert users and didn't find a way)

**Comment by NikunjChavda19** [comment]
@srikanth-sankaran 
Facing this error while opening eclipse and every time while click on project workspace, done google research regarding this also did ask around few people but bot getting any solution, if you have any solution or suggestion! it would help me and would appreciate your support.
Thanks & Regards 

![image](https://github.com/eclipse-egit/egit/assets/149466667/b2bf87e1-6145-4efd-85ef-2f3da2751968)

**Comment by iloveeclipse** [comment]
@NikunjChavda19 : your problem is completely unrelated to this ticket. Please create dedicated issue!

**Comment by NikunjChavda19** [comment]
_**@iloveeclipse**_ : I have already created one.

**Please refer this** : https://github.com/eclipse-egit/egit/issues/1#issue-1969923700

**Comment by iloveeclipse** [comment]
> _**@iloveeclipse**_ : I have already created one.

Fine, so please avoid hijacking other tickets.

**Comment by NikunjChavda19** [comment]
@iloveeclipse Oh ! Can you please let me know the dedicated platform to report the issue ?

**Comment by iloveeclipse** [comment]
> Can you please let me know the dedicated platform to report the issue ?

You **have already reported one**: https://github.com/eclipse-egit/egit/issues/1

**Comment by srikanth-sankaran** [comment]
@NikunjChavda19 - If your issue doesn't get picked up for work in "reasonable time" and it is impacting your work/productivity, you may want to ask in the project mailing lists for help. (See https://projects.eclipse.org/projects/technology.egit/contact). I don't have visibility into the project's plans, priorities and resource constraint etc.

**Comment by NikunjChavda19** [comment]
@srikanth-sankaran : Thank you for the guidance 👍

**Comment by Bananeweizen** [comment]
That's possible since many years: https://git-scm.com/docs/git-config#Documentation/git-config.txt-branchautoSetupRebase (with always). egit will respect that setting. In fact, I have all my company Oomph profiles with that (and some other) config values so everyone gets it without even knowing. Might be good to add it to JDTs Oomph setup too.

**Comment by iloveeclipse** [comment]
Thanks, that works!

**TL;DR**

`branch.autoSetupRebase=always` entry should be added to the git repository properties dialog, the end result will look like:

![image](https://github.com/eclipse-egit/egit/assets/964108/027bcdbc-85ce-4142-8b6d-a94b03492afe)

**Comment by iloveeclipse** [comment]
@merks : for those who use Oomph (I'm thinking on Platform setups), could we automatically change git settings  after clone to have rebase reselected?

**Comment by merks** [comment]
Could you please open an issue and in that issue paste the full config file text of a properly configured repository so that I can be sure I've created the correct structure.  Probably we'd want to do this for all the SDK setup clones, so it doesn't matter in which repo you create the issue...

**Comment by iloveeclipse** [comment]
@merks : https://github.com/eclipse-jdt/eclipse.jdt/issues/79

**Comment by srikanth-sankaran** [comment]
> Thanks, that works!
> 
> **TL;DR**
> 
> `branch.autoSetupRebase=always` entry should be added to the git repository properties dialog

FYI, This works for the  git branch, git switch or git checkout  operations wrapped up by the egit UI. However, if right click on a repo and choose `Pull ...` I still see merge as the prepopulated choice.

**Comment by Bananeweizen** [comment]
Because you are still doing a pull operation. pull is the combination of a fetch and an "integration" operation, and your config setting just changes the behaviour of the integration part to be a rebase instead of a merge. https://www.git-scm.com/docs/git-pull

**Comment by tomaswolf** [comment]
Also set `pull.rebase = true`.

**Comment by merks** [comment]
@tomaswolf 

The existing property looks like this in the setup:

![image](https://github.com/eclipse-egit/egit/assets/208716/65b6b400-eea2-44b2-a6b0-381e5305dc42)

So this new one should be like the selected above?

**Comment by tomaswolf** [comment]
What's that image? An Oomph setup? I don't know how it should look.

If the git config is already set via the Oomph setup, then it's fine. But then I don't understand why srikanth still had the merge option pre-selected.

**Comment by merks** [comment]
Yes, I don't understand either what the purpose of this other option is.   I can set it like above and then it looks like this in the clone:

![image](https://github.com/eclipse-egit/egit/assets/208716/bb6c9dfe-7260-429f-8283-78f7d4d1c5ec)

But what does that do?

When Oomph clones the branch it always sets it to rebase by default, as you can see above.  So a Pull... does this

![image](https://github.com/eclipse-egit/egit/assets/208716/058a60b9-c141-4b2c-839b-ae3f10e749b3)

Before I add the other option, I'd like to understand what it does.

**Comment by Bananeweizen** [comment]
I'm not sure if the "rebase.pull" option needs to be set globally. The behavior is as follows, AFAIK:

* Due to autosetuprebase, every newly created branch will be set up with the rebase=true setting in the git config, therefore all menus and dialogs related to that branch will show the rebase strategy.
* The "Pull..." dialog will take the information stored with the branch to prepopulate the strategy combobox. If there are branches in the local repository that were created before applying the autosetuprebase setting, they don't have individual "branch = true" settings in the git config. When selecting such a branch in the "Pull... dialog", the _repository_ default for rebase/merge will be taken, if it exists. If a repository default doesn't exist, then simply the last used value from the dialog seems to be taken as default.
* The pull.rebase=true suggestion of Tomas adds a repository default, thereby eliminating the case where the dialog uses "whatever was used last time" for branches without an own config.

Does that roughly fit, @tomaswolf or did I miss some aspect? I would think that a freshly created Oomph setup would not need the rebase.pull option, because there are no branches without specific setting, right?

@merks You can experiment on your own by simply deleting all lines belonging to a branch configuration, but not the section of the branch. So for the first branch in this screenshot the dialog will always be rebase, for the second one it will be either repository default or last used: 
![grafik](https://github.com/eclipse-egit/egit/assets/406876/1d5296f2-cb12-4b86-a876-55407d1d369d)
